Add new dom-elements-no-danger-with-children rule (Fixes #710)

petersendidit opened the pull request:
Prevents dangerouslySetInnerHTML and children from being used at the same time (Fixes #710)

ljharb commented on tests/lib/rules/dom-elements-no-danger-with-children.js:
```diff
@@ -0,0 +1,67 @@
+  invalid: [
+    {
+      code: [
+        '<div dangerouslySetInnerHTML={{ __html: "HTML" }}>',
```
Can we also add a test that sets the "children" prop explicitly, in both cases?

petersendidit replied:
Good idea. Added.

ljharb replied:
I see you've added it for `React.createElement`, but not for the jsx example.

petersendidit replied:
Didn't even think about passing a children prop in JSX. Seems just wrong to even do such thing. Will update the rule.

ljharb replied:
It's very wrong but we should still cover it :-)

petersendidit replied:
Added

lencioni replied:
Sounds like another good rule would be to prevent a `children` prop on DOM elements.

ljharb replied:
I'd say on any elements - children should be actual children (in `React.createElement` as well)

lencioni replied:
Ah, right. Good point!

lencioni replied:
#720

lencioni commented on docs/rules/dom-elements-no-danger-with-children.md:
```diff
@@ -0,0 +1,39 @@
+# Prevent problem with children and props.dangerouslySetInnerHTML (dom-elements-no-danger-with-children)
+
+This rule helps prevent problems caused by using children and the dangerouslySetInnerHTML prop at the same time
```
This sentence needs ending punctuation.

Also, it might be worth mentioning that React will issue a warning for this case.

petersendidit commented:
New commit should take care of the comments. Let me know if there is anything else.

yannickcr commented:
Thanks, just a few small things:

* Can you rename the rule to `no-danger-with-children`
* Pass the `recommended` to `false` in the metadata
* Rebase your branch

And then I think we'll be good to go :)

ljharb commented on lib/rules/dom-elements-no-danger-with-children.js:
```diff
@@ -0,0 +1,80 @@
+    docs: {
+      description: 'Report when a DOM element is using both children and dangerouslySetInnerHTML',
+      category: '',
+      recommended: true
```
It should be, though, since it's already a React error.

petersendidit commented:
Renamed and removed it from the recommended rules.

I do think these type of rules that are catching errors that React will throw itself should be recommend ASAP.

lencioni commented:
Agreed. I think we should write them all, put out minor releases, and then
add them all to recommended config in a major release ASAP. That way we can
maximize benefit and minimize pain.

On Sat, Aug 13, 2016, 11:09 AM David Petersen <notifications@github.com>
wrote:

> Renamed and removed it from the recommended rules.
>
> I do think these type of rules that are catching errors that React will
> throw itself should be recommend ASAP.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/yannickcr/eslint-plugin-react/pull/719#issuecomment-239633938>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAL7zovEh2NROCC1sglDA5j8o_GOzYDrks5qfgg_gaJpZM4JVbyE>
> .
>

ljharb commented:
Although, I kind of feel like it's not really a semver-major if React would be throwing the error at runtime anyways.

lencioni commented:
I think it is a warning

On Sat, Aug 13, 2016, 12:59 PM Jordan Harband <notifications@github.com>
wrote:

> Although, I kind of feel like it's not really a semver-major if React
> would be throwing the error at runtime anyways.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/yannickcr/eslint-plugin-react/pull/719#issuecomment-239638717>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAL7zmi84WKrJ8jBOGo4Buksjfh3QeDOks5qfiIQgaJpZM4JVbyE>
> .
>

yannickcr commented:
> I do think these type of rules that are catching errors that React will throw itself should be recommend ASAP.

Yes, it will be added to the recommended config on next major release.

yannickcr commented:
Merges, thanks!
